src/MongoDB.Bson/BsonExtensionMethods.cs

```diff
@@ -43,6 +43,15 @@ public static class BsonExtensionMethods
             BsonSerializationArgs args = default(BsonSerializationArgs)
             )
         {
+            if (args.NominalType == null)
+            {
+                args.NominalType = typeof(TNominalType);
+            }
+            else if (args.NominalType != typeof(TNominalType))
+            {
+                throw new ArgumentException("args.NominalType must be <TNominalType>.", nameof(args));
+            }
+
```

Comment:
It's a bit of a design flaw that nominal type can be specified in two places.

Rather than making breaking changes what we can do is provide the right default value for args.NominalType and validate the value if one was provided.

Reply:
I did want a different error message. But if the duplication bothers you I can reword the error message to apply to both overloads.

Note: the `obj` does not have a nominal type. `obj` has an actual type.

Reply:
Ok, what about:

>NominalType in BsonSerializationArgs must be equal to the passed nominal type.

And then, since the error message will be the same in any case, it's enough to have only one validation for `NominalType`:

        public static byte[] ToBson<TNominalType>(
            this TNominalType obj,
            IBsonSerializer<TNominalType> serializer = null,
            BsonBinaryWriterSettings writerSettings = null,
            Action<BsonSerializationContext.Builder> configurator = null,
            BsonSerializationArgs args = default(BsonSerializationArgs)
            )
        {
            return ToBson(obj, typeof(TNominalType), writerSettings, serializer, configurator, args);
        }

        public static byte[] ToBson(
            this object obj,
            Type nominalType,
            BsonBinaryWriterSettings writerSettings = null,
            IBsonSerializer serializer = null,
            Action<BsonSerializationContext.Builder> configurator = null,
            BsonSerializationArgs args = default(BsonSerializationArgs))
        {
            if (nominalType == null)
            {
                throw new ArgumentNullException("nominalType");
            }

            if (args.NominalType == null)
            {
                args.NominalType = nominalType;
            }
            else if (args.NominalType != nominalType)
            {
                throw new ArgumentException("NominalType in BsonSerializationArgs must be equal to the passed nominal type.", nameof(args));
            }

            if (serializer == null)
            {
                serializer = BsonSerializer.LookupSerializer(nominalType);
            }
            if (serializer.ValueType != nominalType)
            {
                var message = string.Format("Serializer type {0} value type does not match document types {1}.", serializer.GetType().FullName, nominalType.FullName);
                throw new ArgumentException(message, "serializer");
            }

            using (var memoryStream = new MemoryStream())
            {
                using (var bsonWriter = new BsonBinaryWriter(memoryStream, writerSettings ?? BsonBinaryWriterSettings.Defaults))
                {
                    var context = BsonSerializationContext.CreateRoot(bsonWriter, configurator);
                    serializer.Serialize(context, args, obj);
                }
                return memoryStream.ToArray();
            }
        }

Reply:
This is just one file... there are also 6 cases in another file.

So centralizing the logic in the new `SetOrVerifyNominalType` method allows even more code sharing.

I think it is a good thing if the messages are slightly different. That way each version is more accurate.